### DISCUSSION TITLE
Berry talk at conference in AU

### DISCUSSION THREAD

**Discussion by darrylb123** [tasmota/Berry_playground#38] [Get Help - Q&A]
Earlier in the year I submitted a talk proposal for a conference in Australia in April 2024. The proposal was 'Tasmota and Berry, not just for home automation'. The talk was initially rejected but I have now been asked if I could present it. 
My thoughts were an brief overview of ESP32 and Tasmota, then a decent review of using Berry for general purpose embedded tasks. Have there been any presentations in this context that I may use some of the material? I propose to submit the presentation here for someone to use in the future.

**Comment by s-hadinger** [comment]
Awesome. But I'm afraid there is material already available. I know very well the internals of Berry so happy to help if needed.

**Comment by darrylb123** [comment]
Thanks,
When you mention the material available, it that the documentation or other
presentations.
A review of my presentation would be very helpful. I'm certainly no expert.

I propose to:
Short discussion on the esp32
Short history of tasmota
Brief introduction of the rule engine and limitations
Introduce berry
Berry libraries
Tasmota support

Talk length 40min, including questions.
Should be time for a demo or 2

**Reply by s-hadinger** [reply]
Happy to review.

On the Berry side, you can mention that it is heavily inspired from LUA but with a more pythonic syntax. Berry is specifically designed for a small code footprint and low memory usage, while being functional (closures) and object-oriented at the same time.

Berry is VM-based, it compiles to bytecode. Also worth noting that the Berry compiler to bytecode is "single-pass", again to reduce the size of the compiler and reduce memory usage. This limits the amount of optimization it can do.

In term of performance, I remember that berry achieves ~1Mips (1 million simple instruction per second), to have a rough estimate of its compute performance, although most of the time is spent dereferencing attributes names

**Comment by darrylb123** [comment]
Yes, I wanted to implement a PID loop for a little project. Grabbed a
simple python implementation and converted to Berry in about 5 minutes.

**Comment by darrylb123** [comment]
I'd never heard of Berry until I saw the Berry console option in the beta
esp32 version.
Why choose Berry, why not Lua or micro python...
Make an interesting slide...

**Reply by s-hadinger** [reply]
Good question. I spent some time to find this gem.

Berry is "just a library" that you can call from whatever call. MicroPython on the contrary takes control of the entire system and was not really compatible with Tasmota architectuer. LUA was a little too heavy, and the syntax was not so much appreciated (especially index starting at 1 instead of 0 in arrays).

Another contender was https://github.com/cesanta/elk (javascript) but I largely prefrer Python syntax compared to JS.

**Reply by darrylb123** [reply]
I have a first draft [darrylb123](https://github.com/darrylb123) 
/
[EO2024](https://github.com/darrylb123/EO2024) in LibreOffice Impress format. The words are there, just might change the format.

**Reply by s-hadinger** [reply]
Thanks, great content. When mentioning impactful features, you can also mention Matter protocol support. This is an important feature for many users. It also supports ESP8266 devices via a Tasmota Matter Bridge.

**Reply by Jason2866** [reply]
Indeed, nice short Presentation with all mandantory points! As said from Stephan just the already very well Berry Matter implementation is missing.

**Comment by darrylb123** [comment]
Thanks, I have not even looked at Matter. Sounds like I should.

**Comment by darrylb123** [comment]
I see why you bring up Matter. Implemented in Berry. I shall add that. Pushed a new version of the presentation.

**Comment by darrylb123** [comment]
The presentation was delivered today. Good feedback. The final version is [here]
(https://github.com/darrylb123/EO2024/blob/main/EO2024-Tasmota%2BBerry.odp).
The video will be available in a few weeks. I'll post the link when it's complete. Do you want the presentation in the playground somewhere?

**Reply by s-hadinger** [reply]
Awesome. Let's have a discussion with the staff about where to put it

**Reply by s-hadinger** [reply]
We discussed it and decided it would be best within the Tasmota documentation. Is it possible to convert the presentation to a browser friendly format to watch online?

**Comment by darrylb123** [comment]
Just a comment. 
The most surprising thing I learned was that a lot of people thought that Tasmota was just firmware for connecting to Home Assistant. Didn't know about rules and commands let alone Berry. No wonder ESPHome is seen as a better alternative. 
As usual, don't read the doco, just go by what jumps out at you, ie from the GUI. These were seasoned Linux/Open source admins and developers!
Perhaps some thought into including a rule builder into the UI, even possibly a button that downloads a Berry program that can be removed when the rules are all built.

**Reply by s-hadinger** [reply]
Good suggestion, but we are lacking front-end developers in the Tasmota project to do all the js/css magic for such an editor. If you know volunteers...

**Comment by darrylb123** [comment]
Yes, should be able to. Leave it with me.

**Comment by darrylb123** [comment]
After nearly a year, the organisers have uploaded the video to YouTube : https://youtu.be/S-oxBUg-FAQ?si=2kbhRzgvqDd7CGU2
I have exported the presentation as PDF and html from Libre Office Impress here: https://github.com/darrylb123/EO2024

Let me know how to proceed from here?
